cargo-nextest/src/dispatch.rs

```diff
@@ -597,38 +598,53 @@ impl TestBuildFilter {
         &self,
         run_ignored: &mut Option<RunIgnored>,
         patterns: &mut Vec<String>,
+        filter_exprs: &mut Vec<FilteringExpr>,
+        graph: &PackageGraph,
     ) -> Result<()> {
         let mut ignore_filters = Vec::new();
         let mut read_trailing_filters = false;
-
-        let mut skip_exact = Vec::new();
         let mut unsupported_args = Vec::new();
-
-        patterns.extend(
-            self.filters
-                .iter()
-                .filter(|&s| {
-                    if read_trailing_filters || !s.starts_with('-') {
-                        true
-                    } else if s == "--include-ignored" {
-                        ignore_filters.push((s.clone(), RunIgnored::All));
-                        false
-                    } else if s == "--ignored" {
-                        ignore_filters.push((s.clone(), RunIgnored::IgnoredOnly));
-                        false
-                    } else if s == "--" {
-                        read_trailing_filters = true;
-                        false
-                    } else if s == "--skip" || s == "--exact" {
-                        skip_exact.push(s.clone());
-                        false
-                    } else {
-                        unsupported_args.push(s.clone());
-                        true
-                    }
-                })
-                .cloned(),
-        );
+        let mut filter_expr_parse_errors = Vec::new();
+
+        let mut it = self.filters.iter();
+        while let Some(arg) = it.next() {
+            if read_trailing_filters || !arg.starts_with('-') {
+                patterns.push(arg.clone());
+            } else if arg == "--include-ignored" {
+                ignore_filters.push((arg.clone(), RunIgnored::All));
+            } else if arg == "--ignored" {
+                ignore_filters.push((arg.clone(), RunIgnored::IgnoredOnly));
+            } else if arg == "--" {
+                read_trailing_filters = true;
+            } else if arg == "--skip" {
+                let skip_arg = it.next().ok_or_else(|| {
+                    ExpectedError::test_binary_args_parse_error(
+                        "missing required argument",
+                        vec![arg.clone()],
+                    )
+                })?;
+
+                match FilteringExpr::parse(format!("!test({skip_arg})"), graph) {
```

Comment:
I don't think this is quite right set-logic wise. The general selection logic here is that the set of tests run is: (UNION of all tests selected by regular arguments) INTERSECTION (UNION of all of tests selected by filtersets).

Based on this, here's what I think the behavior should be:

* `--exact` should be treated as part of the regular arguments (`patterns`).

  For example: `cargo nextest run -- a b -E 'test(c)' -E 'test(=d)' --exact e` should select tests that (have "a" in their name UNION have "b" in their name UNION named exactly "e") INTERSECTION (have "c" in their name UNION named exactly "d").

  This case fits nicely into the existing logic for regular arguments (but it should be treated as part of regular arguments, not filtersets).

* `--skip` should be applied at the end after everything.

  For example, `cargo nextest run -- a b -E 'test(c)' -E 'test(=d)' --skip e` should select tests that (have "a" in their name UNION have "b" in their name) INTERSECTION (have "c" in their name UNION named exactly "d") INTERSECTION (not have "e" in their name). This needs to be handled separately, as a third thing to intersect against.

While it's definitely possible to use filtersets to implement this logic, I think the requirement to pass in a graph here makes it somewhat problematic so I'd prefer just using regular arguments.

So here's what I would do:

1. Expand `patterns` so that rather than being a `Vec<String>` it is a `Vec<Pattern>`, where `Pattern` is an enum with `Contains` and `Exact` variants.
2. Also add a separate `skip_patterns`.
3. Add some set logic tests for all of these cases (I can help out with this if you get stuck, just doing 1 and 2 is okay and I can pick up 3.)

Reply:
I force-pushed an incomplete version of this. I'm not clear on why `exact` would be added to the existing `patterns` and `skip` would be its own list. It seems to me like the different sorts of patterns need to be picked apart in `TestFilterBuilder::new` anyways, so might as well use one `Vec` for each (potentially bundled into a `struct`), or put them all in one `Vec`?

I also just saw this needs rebasing, will do that when the above question is answered, alongside fixing the missing docs and filtering logic.

Reply:
Sure, you can use a separate Vec for exact patterns. The important thing is to get the set logic right.

Reply:
Having written the code -- agreed, a separate Vec is better. Thanks! :D